File: services/drishtix-app/src/main/java/com/drishtix/model/CameraSource.java
package com.drishtix.model;

import java.time.LocalDateTime;

/**
 * Entity representing a camera source.
 * Maps to the {@code camera_sources} table in the database.
 */
public class CameraSource {

    private int cameraId;
    private String cameraName;
    private String sourceUri;
    private boolean active;
    private LocalDateTime createdAt;

    public CameraSource() {
    }

    public CameraSource(String cameraName, String sourceUri) {
        this.cameraName = cameraName;
        this.sourceUri = sourceUri;
        this.active = true;
    }

    public int getCameraId() {
        return cameraId;
    }

    public void setCameraId(int cameraId) {
        this.cameraId = cameraId;
    }

    public String getCameraName() {
        return cameraName;
    }

    public void setCameraName(String cameraName) {
        this.cameraName = cameraName;
    }

    public String getSourceUri() {
        return sourceUri;
    }

    public void setSourceUri(String sourceUri) {
        this.sourceUri = sourceUri;
    }

    public boolean isActive() {
        return active;
    }

    public void setActive(boolean active) {
        this.active = active;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }

    /**
     * Returns the source as an integer device index if it's numeric,
     * otherwise returns -1 (indicating an RTSP/URL source).
     */
    public int getDeviceIndex() {
        try {
            return Integer.parseInt(sourceUri.trim());
        } catch (NumberFormatException e) {
            return -1;
        }
    }

    @Override
    public String toString() {
        return cameraName + " (" + sourceUri + ")";
    }
}

File: services/drishtix-py/drishtix/models/camera_source.py
"""
DrishtiX v4.0 — CameraSource ORM model.

Migrated from: com.drishtix.model.CameraSource (Java)
MongoDB collection: 'camera_sources' → SQLite table: 'camera_source'
"""

from datetime import datetime
from typing import Optional

from sqlalchemy import Boolean, DateTime, String, func
from sqlalchemy.orm import Mapped, mapped_column

from drishtix.models.base import Base


class CameraSource(Base):
    """
    Camera source entity.

    Represents a connected camera device (USB webcam, integrated laptop camera)
    or a remote video stream (RTSP/HTTP URL). The source_uri field accepts
    either an integer device index (e.g., '0') or a stream URL.
    """

    __tablename__ = "camera_source"

    camera_id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    camera_name: Mapped[str] = mapped_column(String(255), nullable=False)
    source_uri: Mapped[str] = mapped_column(
        String(500),
        nullable=False,
        comment="Device index (int) or RTSP/HTTP URL (string)",
    )
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime, default=func.now(), nullable=False
    )

    @property
    def device_index(self) -> int:
        """
        Parse source_uri as an integer device index.

        Returns:
            The device index if source_uri is numeric, otherwise -1
            (indicating an RTSP/URL source).
        """
        try:
            return int(self.source_uri.strip())
        except (ValueError, AttributeError):
            return -1

    @property
    def is_stream(self) -> bool:
        """True if this source is a network stream (not a local device)."""
        return self.device_index == -1

    def __repr__(self) -> str:
        return f"CameraSource(id={self.camera_id}, name='{self.camera_name}', uri='{self.source_uri}')"
